src/transport/tests/TestTCP.cpp

```diff
@@ -0,0 +1,216 @@
+
+using TCPImpl = Transport::TCP<kMaxTcpActiveConnectionCount, kMaxTcpPendingPackets>;
+
+constexpr NodeId kSourceNodeId      = 123654;
+constexpr NodeId kDestinationNodeId = 111222333;
+constexpr uint32_t kMessageId       = 18;
+
+using TestContext = chip::Test::IOContext;
+TestContext sContext;
+
+const char PAYLOAD[]        = "Hello!";
+int ReceiveHandlerCallCount = 0;
+
+void MessageReceiveHandler(MessageHeader & header, const Transport::PeerAddress & source, System::PacketBuffer * msgBuf,
+                           nlTestSuite * inSuite)
+{
+    NL_TEST_ASSERT(inSuite, header.GetSourceNodeId() == Optional<NodeId>::Value(kSourceNodeId));
+    NL_TEST_ASSERT(inSuite, header.GetDestinationNodeId() == Optional<NodeId>::Value(kDestinationNodeId));
+    NL_TEST_ASSERT(inSuite, header.GetMessageId() == kMessageId);
+
+    size_t data_len = msgBuf->DataLength();
+    int compare     = memcmp(msgBuf->Start(), PAYLOAD, data_len);
+    NL_TEST_ASSERT(inSuite, compare == 0);
+
+    ReceiveHandlerCallCount++;
+}
+
+/////////////////////////// Init test
+
+void CheckSimpleInitTest(nlTestSuite * inSuite, void * inContext, Inet::IPAddressType type)
+{
+    TestContext & ctx = *reinterpret_cast<TestContext *>(inContext);
+
+    TCPImpl tcp;
+
+    CHIP_ERROR err = tcp.Init(Transport::TcpListenParameters(&ctx.GetInetLayer()).SetAddressType(type));
+
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+}
+
+#if INET_CONFIG_ENABLE_IPV4
+void CheckSimpleInitTest4(nlTestSuite * inSuite, void * inContext)
+{
+    CheckSimpleInitTest(inSuite, inContext, kIPAddressType_IPv4);
+}
+#endif
+
+void CheckSimpleInitTest6(nlTestSuite * inSuite, void * inContext)
+{
+    CheckSimpleInitTest(inSuite, inContext, kIPAddressType_IPv6);
+}
+
+/////////////////////////// Messaging test
+
+void CheckMessageTest(nlTestSuite * inSuite, void * inContext, const IPAddress & addr)
+{
+    TestContext & ctx = *reinterpret_cast<TestContext *>(inContext);
+
+    size_t payload_len = sizeof(PAYLOAD);
+
+    chip::System::PacketBuffer * buffer = chip::System::PacketBuffer::NewWithAvailableSize(payload_len);
+    memmove(buffer->Start(), PAYLOAD, payload_len);
+    buffer->SetDataLength(payload_len);
+
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    TCPImpl tcp;
+
+    err = tcp.Init(Transport::TcpListenParameters(&ctx.GetInetLayer()).SetAddressType(addr.Type()));
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    tcp.SetMessageReceiveHandler(MessageReceiveHandler, inSuite);
+    ReceiveHandlerCallCount = 0;
+
+    MessageHeader header;
+    header.SetSourceNodeId(kSourceNodeId).SetDestinationNodeId(kDestinationNodeId).SetMessageId(kMessageId);
+
+    // Should be able to send a message to itself by just calling send.
+    err = tcp.SendMessage(header, Transport::PeerAddress::TCP(addr), buffer);
+    if (err == System::MapErrorPOSIX(EADDRNOTAVAIL))
+    {
+        // TODO: the underlying system does not support IPV6. This early return should
```

Comment:
Please file an issue tracking this.

Reply:
This is a copy and paste from TestUDP ... at that time we had errors in CI, which should already have tracking issues.
I will update the wording.

Reply:
Added an issue after all, since I believe the original issue for CI is already addressed in a curent PR which likely will not touch TestTCP/UDP.

Opened issue is #2698 and is now referenced in the TODO.